File: examples/ts/bsv21-token/BSV21Token.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract, ALICE, signTestMessage } from 'runar-testing';
import { compile } from 'runar-compiler';
import type { RunarArtifact } from 'runar-ir-schema';
import { RunarContract } from '../../../packages/runar-sdk/src/contract.js';
import {
  parseInscriptionEnvelope,
} from '../../../packages/runar-sdk/src/ordinals/envelope.js';
import { BSV21 } from '../../../packages/runar-sdk/src/ordinals/bsv20.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'BSV21Token.runar.ts'), 'utf8');

const SIG = signTestMessage(ALICE.privKey);

/** Convert hex to UTF-8 string. */
function hexToUtf8(hex: string): string {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.slice(i, i + 2), 16);
  }
  return new TextDecoder().decode(bytes);
}

function compileContract(): RunarArtifact {
  const result = compile(source, { fileName: 'BSV21Token.runar.ts' });
  if (!result.success || !result.artifact) {
    const errors = result.diagnostics
      .filter((d) => d.severity === 'error')
      .map((d) => d.message)
      .join('\n  ');
    throw new Error(`BSV21Token compile failed:\n  ${errors}`);
  }
  return result.artifact;
}

describe('BSV21Token', () => {
  describe('business logic', () => {
    it('accepts a valid unlock', () => {
      const contract = TestContract.fromSource(source, { pubKeyHash: ALICE.pubKeyHash });
      const result = contract.call('unlock', { sig: SIG, pubKey: ALICE.pubKey });
      expect(result.success).toBe(true);
    });
  });

  describe('compilation', () => {
    it('compiles to a valid artifact', () => {
      const artifact = compileContract();
      expect(artifact.contractName).toBe('BSV21Token');
    });
  });

  describe('BSV-21 deploy+mint inscription', () => {
    it('creates a deploy+mint inscription with all fields', () => {
      const artifact = compileContract();
      const inscription = BSV21.deployMint({
        amt: '1000000',
        dec: '18',
        sym: 'RNR',
        icon: 'b61b0172d95e266c18aea0c624db987e971a5d6d4ebc2aaed85da4642d635735_0',
      });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);

      expect(parsed).not.toBeNull();
      expect(parsed!.contentType).toBe('application/bsv-20');

      const json = JSON.parse(hexToUtf8(parsed!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('deploy+mint');
      expect(json.amt).toBe('1000000');
      expect(json.dec).toBe('18');
      expect(json.sym).toBe('RNR');
      expect(json.icon).toBe('b61b0172d95e266c18aea0c624db987e971a5d6d4ebc2aaed85da4642d635735_0');
    });

    it('creates a deploy+mint inscription with minimal fields', () => {
      const artifact = compileContract();
      const inscription = BSV21.deployMint({ amt: '500' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);

      const json = JSON.parse(hexToUtf8(parsed!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('deploy+mint');
      expect(json.amt).toBe('500');
      expect(json.dec).toBeUndefined();
      expect(json.sym).toBeUndefined();
    });
  });

  describe('BSV-21 transfer inscription', () => {
    it('creates a transfer inscription with token ID', () => {
      const artifact = compileContract();
      const tokenId = '3b313338fa0555aebeaf91d8db1ffebd74773c67c8ad5181ff3d3f51e21e0000_1';
      const inscription = BSV21.transfer({ id: tokenId, amt: '100' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);

      expect(parsed).not.toBeNull();
      expect(parsed!.contentType).toBe('application/bsv-20');

      const json = JSON.parse(hexToUtf8(parsed!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('transfer');
      expect(json.id).toBe(tokenId);
      expect(json.amt).toBe('100');
    });
  });

  describe('round-trip', () => {
    it('deploy+mint inscription survives fromUtxo round-trip', () => {
      const artifact = compileContract();
      const inscription = BSV21.deployMint({ amt: '1000000', sym: 'RNR' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();

      const reconnected = RunarContract.fromUtxo(artifact, {
        txid: '00'.repeat(32),
        outputIndex: 0,
        satoshis: 1,
        script: lockingScript,
      });

      expect(reconnected.inscription).not.toBeNull();
      expect(reconnected.inscription!.contentType).toBe('application/bsv-20');

      const json = JSON.parse(hexToUtf8(reconnected.inscription!.data));
      expect(json.p).toBe('bsv-20');
      expect(json.op).toBe('deploy+mint');
      expect(json.amt).toBe('1000000');
      expect(json.sym).toBe('RNR');
    });

    it('transfer inscription survives fromUtxo round-trip', () => {
      const artifact = compileContract();
      const tokenId = 'abc123_0';
      const inscription = BSV21.transfer({ id: tokenId, amt: '50' });
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);
      contract.withInscription(inscription);

      const lockingScript = contract.getLockingScript();

      const reconnected = RunarContract.fromUtxo(artifact, {
        txid: '00'.repeat(32),
        outputIndex: 0,
        satoshis: 1,
        script: lockingScript,
      });

      const json = JSON.parse(hexToUtf8(reconnected.inscription!.data));
      expect(json.op).toBe('transfer');
      expect(json.id).toBe(tokenId);
      expect(json.amt).toBe('50');
    });
  });
});

File: examples/rust/bsv21-token/BSV21Token_test.rs
#[path = "BSV21Token.runar.rs"]
mod contract;

use contract::*;
use runar::prelude::*;
use runar::sdk::{
    ordinals::{bsv21_deploy_mint, bsv21_transfer, parse_inscription_envelope},
    RunarArtifact, RunarContract, SdkValue, Utxo,
};
use runar_compiler_rust::compile_from_source_str;

const SOURCE: &str = include_str!("BSV21Token.runar.rs");

/// Convert hex to a UTF-8 string.
fn hex_to_utf8(hex: &str) -> String {
    let bytes: Vec<u8> = (0..hex.len())
        .step_by(2)
        .filter_map(|i| u8::from_str_radix(&hex[i..i + 2], 16).ok())
        .collect();
    String::from_utf8_lossy(&bytes).into_owned()
}

/// Compile the contract source and deserialize into the SDK's RunarArtifact.
fn compile_artifact() -> RunarArtifact {
    let compiler_artifact = compile_from_source_str(SOURCE, Some("BSV21Token.runar.rs"))
        .expect("BSV21Token must compile");
    let json = serde_json::to_string(&compiler_artifact).expect("artifact JSON");
    serde_json::from_str(&json).expect("SDK artifact deserialize")
}

fn constructor_args() -> Vec<SdkValue> {
    vec![SdkValue::Bytes(
        ALICE
            .pub_key_hash
            .iter()
            .map(|b| format!("{:02x}", b))
            .collect(),
    )]
}

// ---------------------------------------------------------------------------
// Business logic
// ---------------------------------------------------------------------------

#[test]
fn test_unlock_accepts_matching_key() {
    let pk = ALICE.pub_key.to_vec();
    let c = BSV21Token { pub_key_hash: hash160(&pk) };
    c.unlock(&ALICE.sign_test_message(), &pk);
}

#[test]
#[should_panic]
fn test_unlock_rejects_wrong_key() {
    let pk = ALICE.pub_key.to_vec();
    let wrong_pk = BOB.pub_key.to_vec();
    let c = BSV21Token { pub_key_hash: hash160(&pk) };
    c.unlock(&BOB.sign_test_message(), &wrong_pk);
}

// ---------------------------------------------------------------------------
// Compile check (Rúnar frontend: parse -> validate -> typecheck)
// ---------------------------------------------------------------------------

#[test]
fn test_compile() {
    runar::compile_check(SOURCE, "BSV21Token.runar.rs").unwrap();
}

#[test]
fn test_compiles_to_artifact() {
    let artifact = compile_artifact();
    assert_eq!(artifact.contract_name, "BSV21Token");
}

// ---------------------------------------------------------------------------
// BSV-21 deploy+mint inscription
// ---------------------------------------------------------------------------

#[test]
fn test_deploy_mint_inscription_with_all_fields() {
    let artifact = compile_artifact();
    let inscription = bsv21_deploy_mint(
        "1000000",
        Some("18"),
        Some("RNR"),
        Some("b61b0172d95e266c18aea0c624db987e971a5d6d4ebc2aaed85da4642d635735_0"),
    );
    let mut contract = RunarContract::new(artifact, constructor_args());
    contract.with_inscription(inscription);

    let locking_script = contract.get_locking_script();
    let parsed = parse_inscription_envelope(&locking_script)
        .expect("envelope must parse from locking script");

    assert_eq!(parsed.content_type, "application/bsv-20");
    let json = hex_to_utf8(&parsed.data);
    assert!(json.contains(r#""p":"bsv-20""#));
    assert!(json.contains(r#""op":"deploy+mint""#));
    assert!(json.contains(r#""amt":"1000000""#));
    assert!(json.contains(r#""dec":"18""#));
    assert!(json.contains(r#""sym":"RNR""#));
    assert!(json.contains(
        r#""icon":"b61b0172d95e266c18aea0c624db987e971a5d6d4ebc2aaed85da4642d635735_0""#
    ));
}

#[test]
fn test_deploy_mint_inscription_minimal() {
    let artifact = compile_artifact();
    let inscription = bsv21_deploy_mint("500", None, None, None);
    let mut contract = RunarContract::new(artifact, constructor_args());
    contract.with_inscription(inscription);

    let locking_script = contract.get_locking_script();
    let parsed = parse_inscription_envelope(&locking_script).unwrap();
    let json = hex_to_utf8(&parsed.data);

    assert!(json.contains(r#""p":"bsv-20""#));
    assert!(json.contains(r#""op":"deploy+mint""#));
    assert!(json.contains(r#""amt":"500""#));
    assert!(!json.contains("\"dec\""));
    assert!(!json.contains("\"sym\""));
    assert!(!json.contains("\"icon\""));
}

// ---------------------------------------------------------------------------
// BSV-21 transfer inscription
// ---------------------------------------------------------------------------

#[test]
fn test_transfer_inscription_with_token_id() {
    let artifact = compile_artifact();
    let token_id = "3b313338fa0555aebeaf91d8db1ffebd74773c67c8ad5181ff3d3f51e21e0000_1";
    let inscription = bsv21_transfer(token_id, "100");
    let mut contract = RunarContract::new(artifact, constructor_args());
    contract.with_inscription(inscription);

    let locking_script = contract.get_locking_script();
    let parsed = parse_inscription_envelope(&locking_script).unwrap();

    assert_eq!(parsed.content_type, "application/bsv-20");
    let json = hex_to_utf8(&parsed.data);
    assert!(json.contains(r#""p":"bsv-20""#));
    assert!(json.contains(r#""op":"transfer""#));
    assert!(json.contains(&format!(r#""id":"{}""#, token_id)));
    assert!(json.contains(r#""amt":"100""#));
}

// ---------------------------------------------------------------------------
// Round-trip via from_utxo
// ---------------------------------------------------------------------------

#[test]
fn test_deploy_mint_survives_from_utxo_round_trip() {
    let artifact = compile_artifact();
    let inscription = bsv21_deploy_mint("1000000", None, Some("RNR"), None);
    let mut contract = RunarContract::new(artifact.clone(), constructor_args());
    contract.with_inscription(inscription);

    let locking_script = contract.get_locking_script();

    let reconnected = RunarContract::from_utxo(
        artifact,
        &Utxo {
            txid: "00".repeat(32),
            output_index: 0,
            satoshis: 1,
            script: locking_script,
        },
    );

    let insc = reconnected
        .inscription()
        .expect("reconnected contract must carry the inscription");
    assert_eq!(insc.content_type, "application/bsv-20");

    let json = hex_to_utf8(&insc.data);
    assert!(json.contains(r#""p":"bsv-20""#));
    assert!(json.contains(r#""op":"deploy+mint""#));
    assert!(json.contains(r#""amt":"1000000""#));
    assert!(json.contains(r#""sym":"RNR""#));
}

#[test]
fn test_transfer_survives_from_utxo_round_trip() {
    let artifact = compile_artifact();
    let token_id = "abc123_0";
    let inscription = bsv21_transfer(token_id, "50");
    let mut contract = RunarContract::new(artifact.clone(), constructor_args());
    contract.with_inscription(inscription);

    let locking_script = contract.get_locking_script();

    let reconnected = RunarContract::from_utxo(
        artifact,
        &Utxo {
            txid: "00".repeat(32),
            output_index: 0,
            satoshis: 1,
            script: locking_script,
        },
    );

    let insc = reconnected
        .inscription()
        .expect("reconnected contract must carry the inscription");
    let json = hex_to_utf8(&insc.data);
    assert!(json.contains(r#""op":"transfer""#));
    assert!(json.contains(&format!(r#""id":"{}""#, token_id)));
    assert!(json.contains(r#""amt":"50""#));
}
